server/channels/app/post.go

```diff
@@ -2275,3 +2295,31 @@ func includeEmbedsAndImages(a *App, c request.CTX, topThreadList *model.TopThrea
 func (a *App) isPostPriorityEnabled() bool {
 	return a.Config().FeatureFlags.PostPriority && *a.Config().ServiceSettings.PostPriority
 }
+
+func (a *App) applyPostsWillBeConsumedHook(posts map[string]*model.Post) {
+	postsSlice := make([]*model.Post, 0, len(posts))
+
+	for _, post := range posts {
+		postsSlice = append(postsSlice, post.ForPlugin())
+	}
```

Comment:
It's unfortunate that this occurs even if there is no plugin implementing this hook. Perhaps we should extend the plugin environment to allow preemptive querying for a case like this?

Reply:
I agree with Jesse that adding this to every function retrieving posts, even if no plugin will use it, seems dangerous performance-wise. Maybe we could extend the `HooksManager` with something like `IsHookUsed(hookId int) bool`. Ideally, we would cache that info (maybe maintaining a set containing the used hooks when adding/removing a product?) so that asking for it is as cheap as possible. Then, it would be a matter of wrapping the new logic in this PR inside that `IsHookUsed`.
It's true that this introduces a change in how plugins have historically worked, so maybe we want to give it a second thought, but I'd be more comfortable merging something like that.

Reply:
Could we wrap this in a feature flag? Otherwise we can start taking a look at extending the HooksManager.

Reply:
Yep, a feature flag would be a suitable way to constrain this! I like it :)

Reply:
If there's a plan to eventually remove the feature flag (via the `HooksManager` extension or through any other solution we can come up with), I'm ok with it :)

Reply:
Added the feature flags to the functions to avoid any processing if it's not on.

Reply:
We plan to take a look at extending HooksManager or alternative solutions and have a PR up in the next 3 months. If there's a preferred solution, please let us know.

Reply:
That plan sounds good. Extending HooksManager seems like a good idea to me, but feel free to ask any questions or share any plan in Community if you want to discuss the implementation before opening a PR.